<h2 style="color:#00008B"> GOOGLE PLAY - DATA CLEANING </h2>

<p style="color:green"> This project utilizes the Google Play Store dataset obtained from Kaggle, comprising over 2 million rows. The dataset contains several key columns that provide valuable insights into app characteristics and user interactions.<br><br>
<b>Ad Supported:</b> Indicates if the app is ad-supported. <br><br>
<b>App Id:</b> Unique identifier for each app. <br><br>
<b>App Name:</b> The name of the app. <br><br>
<b>Category:</b> The category to which the app belongs (e.g., Games, Productivity). <br><br>
<b>Content Rating:</b> Age rating for the app's content. <br><br>
<b>Currency:</b> Currency used for pricing. <br><br>
<b>Developer Email:</b> Email contact for the developer. <br><br>
<b>Developer Id:</b> Unique identifier for the developer. <br><br>
<b>Developer Website:</b> Website of the app developer. <br><br>
<b>Editors Choice:</b> Indicates if the app is an editor's choice. <br><br>
<b>Free:</b> Indicates if the app is free to download. <br><br>
<b>In-App Purchases:</b> Indicates if the app offers in-app purchases. <br><br>
<b>Installs:</b> Total number of installs for the app. <br><br>
<b>Last Updated:</b> Date when the app was last updated. <br><br>
<b>Maximum Installs:</b> The highest number of installs recorded. <br><br>
<b>Minimum Android:</b> Minimum Android version required to run the app. <br><br>
<b>Minimum Installs:</b> Minimum number of installs recorded. <br><br>
<b>Price:</b> The price of the app. <br><br>
<b>Privacy Policy:</b> Link to the app's privacy policy. <br><br>
<b>Rating:</b> Average user rating for the app. <br><br>
<b>Released:</b> Release date of the app. <br><br>
<b>Scraped Time:</b> Timestamp of when the data was scraped. <br><br>
<b>Size:</b> Size of the app in MB. <br>
</p>

<h4 style="color:#00008B"> Importing Libraries </h4>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Create a DataFrame from the Google Play Store CSV file
data = pd.read_csv("Google-Playstore.csv")
data.head(2)

App Name                      App Id   Category  Rating  \
0              Gakondo         com.ishakwe.gakondo  Adventure     0.0   
1  Ampere Battery Info  com.webserveis.batteryinfo      Tools     4.4   

   Rating Count Installs  Minimum Installs  Maximum Installs  Free  Price  \
0           0.0      10+              10.0                15  True    0.0   
1          64.0   5,000+            5000.0              7662  True    0.0   

   ...                Developer Website         Developer Email      Released  \
0  ...      https://beniyizibyose.tk/#/  jean21101999@gmail.com  Feb 26, 2020   
1  ...  https://webserveis.netlify.app/    webserveis@gmail.com  May 21, 2020   

   Last Updated Content Rating  \
0  Feb 26, 2020       Everyone   
1  May 06, 2021       Everyone   

                                      Privacy Policy Ad Supported  \
0                 https://beniyizibyose.tk/projects/        False   
1  https://dev4phones.wordpress.com/licencia-de-uso/         True   

  In App Purchases Editors Choice         Scraped Time  
0            False          False  2021-06-15 20:19:35  
1            False          False  2021-06-15 20:19:35  

[2 rows x 24 columns]

<h2 style="color:#00008B">Data Cleaning </h2>

In [3]:
# List the columns in the DataFrame
data.columns

Index(['App Name', 'App Id', 'Category', 'Rating', 'Rating Count', 'Installs',
       'Minimum Installs', 'Maximum Installs', 'Free', 'Price', 'Currency',
       'Size', 'Minimum Android', 'Developer Id', 'Developer Website',
       'Developer Email', 'Released', 'Last Updated', 'Content Rating',
       'Privacy Policy', 'Ad Supported', 'In App Purchases', 'Editors Choice',
       'Scraped Time'],
      dtype='object')

<p style='color:green'> The following columns are irrelevant to our analysis and will be excluded: 'Ad Supported', 'App Id', 'Currency', 'Developer Email', 'Developer Website', and 'Privacy Policy'. </p>

In [4]:
# Removing columns that are not relevant for analysis
data.drop(columns=['Ad Supported', 'App Id', 'Currency', 'Developer Email', 'Developer Website','Privacy Policy'], inplace=True)

In [5]:
# no. of rows and columns checking
data.shape

(2312944, 18)

In [6]:
# displaying sum of null
data.isnull().sum()

App Name                5
Category                0
Rating              22883
Rating Count        22883
Installs              107
Minimum Installs      107
Maximum Installs        0
Free                    0
Price                   0
Size                  196
Minimum Android      6530
Developer Id           33
Released            71053
Last Updated            0
Content Rating          0
In App Purchases        0
Editors Choice          0
Scraped Time            0
dtype: int64

<p style='color:green'> The dataset contains missing values across several key features. Specifically, both Rating and Rating Count have 22,883 missing values, which accounts for 0.989% of the total entries. Additionally, there are 107 missing values for both Installs and Minimum Installs, representing 0.0046%. The Size field is missing 196 values (0.0085%), while Minimum Android Version is missing 6,530 values, or 0.2824% of the dataset. Furthermore, 33 entries (0.0014%) are missing for the Developer ID field, and the Release Date is missing 71,053 values, which is 3.072% of the total.</p>

In [7]:
# Fill missing values with "Unknown"
columns_to_replace = ['Installs', 'Minimum Installs', 'Size', 'Minimum Android', 'Developer Id', 'Released','App Name']
data[columns_to_replace] = data[columns_to_replace].fillna("Unknown")

In [8]:
# displaying sum of null
data.isnull().sum()

App Name                0
Category                0
Rating              22883
Rating Count        22883
Installs                0
Minimum Installs        0
Maximum Installs        0
Free                    0
Price                   0
Size                    0
Minimum Android         0
Developer Id            0
Released                0
Last Updated            0
Content Rating          0
In App Purchases        0
Editors Choice          0
Scraped Time            0
dtype: int64

In [9]:
# Drop NA values in Rating and Rating Count
data = data.dropna(subset=['Rating', 'Rating Count'])

In [10]:
# displaying sum of null
data.isnull().sum()

App Name            0
Category            0
Rating              0
Rating Count        0
Installs            0
Minimum Installs    0
Maximum Installs    0
Free                0
Price               0
Size                0
Minimum Android     0
Developer Id        0
Released            0
Last Updated        0
Content Rating      0
In App Purchases    0
Editors Choice      0
Scraped Time        0
dtype: int64

<h2 style="color:#00008B"> Feature Engineering </h2>

<p style="color:green"> Feature engineering is the process of using domain knowledge to create new input features or modify existing ones from raw data to make our dataset columns more meaningful for our analysis.
We will perform feature engineering on few columns like : Size, Minimum Android, Installs, Maximum and Minimum Installs, Content Rating, Rating, In App Purchases, Rating Count, Price</p>

<h4 style='color:green'> ~ Size </h4>

In [11]:
def convert_size_to_mb(size):
    
    #  'Varies with device' and 'Unknown'
    if size == 'Varies with device' or size == 'Unknown':
        return "Unknown" 
    
    # Remove commas from numbers
    size = size.replace(',', '')
    
    # Check the last character for unit type
    unit = size[-1]
    value = float(size[:-1])  # Get the numerical part before the unit
    
    if unit == 'k':  
        mb_value = value / 1024
        return max(1, int(mb_value))  # If less than 1MB, round up to 1
    elif unit == 'M':  
        return int(value)
    elif unit == 'G':  
        return int(value * 1024)
    else:
        return "Unknown"  

data['Size'] = data['Size'].apply(convert_size_to_mb)

<h4 style='color:green'> ~ Minimum Android </h4>

In [12]:
def categorize_android_version(version):
    # Handle 'Unknown' and 'Varies with device'
    if version == 'Unknown':
        return 'Unknown'
    elif version == 'Varies with device':
        return 'Varies with device'
    
    # Extract the first part of the version (major version)
    try:
        major_version = float(version.split('.')[0])  # Get the version before dot (e.g., '5.4.1' -> 5.0)
    except ValueError:
        return 'Invalid' 
        
    # Categorize versions
    if major_version < 5:
        return 'Below 5.0'
    elif 5 <= major_version < 8:
        return '5.0 to 8.0'
    else:
        return '8.0 and up'

data['Minimum Android'] = data['Minimum Android'].apply(categorize_android_version)

In [13]:
data['Minimum Android']

0          5.0 to 8.0
1          5.0 to 8.0
2           Below 5.0
3           Below 5.0
4           Below 5.0
              ...    
2312939     Below 5.0
2312940     Below 5.0
2312941    5.0 to 8.0
2312942    5.0 to 8.0
2312943    5.0 to 8.0
Name: Minimum Android, Length: 2290061, dtype: object

<h4 style='color:green'> ~ Installs </h4>

In [14]:
def categorize_installs(install):
    if install.strip() == 'Unknown':
        return 'Unknown'
    
    # Remove commas and '+' sign, and convert to integer
    install_cleaned = install.replace(',', '').replace('+', '').strip()
    
    try:
        install_number = int(install_cleaned)
    except ValueError:
        return 'Invalid'  
    
    # Categorize based on install ranges
    if install_number < 10000:
        return '0 to 10,000'
    elif install_number < 100000:
        return '10,001 to 100,000'
    elif install_number < 1000000:
        return '100,001 to 1,000,000'
    else:
        return '1,000,001 and above'

data['Installs'] = data['Installs'].apply(categorize_installs)

In [15]:
data['Installs']

0                   0 to 10,000
1                   0 to 10,000
2                   0 to 10,000
3                   0 to 10,000
4                   0 to 10,000
                   ...         
2312939    100,001 to 1,000,000
2312940             0 to 10,000
2312941             0 to 10,000
2312942             0 to 10,000
2312943             0 to 10,000
Name: Installs, Length: 2290061, dtype: object

<h4 style='color:green'> ~ Minimum and Maximum Installs </h4>

In [16]:
# Convert Minimum Installs to numeric
data['Minimum Installs'] = pd.to_numeric(data['Minimum Installs'], errors='coerce')

# Remove rows where Minimum Installs is 0
data = data[data['Minimum Installs'] > 0]

# Categorize Minimum Installs
def categorize_minimum_installs(min_installs):
    if min_installs < 1000:
        return 'Low'
    elif min_installs < 10000:
        return 'Medium'
    else:
        return 'High'

data['Minimum Installs'] = data['Minimum Installs'].apply(categorize_minimum_installs)

# Categorize Maximum Installs
def categorize_maximum_installs(max_installs):
    if max_installs < 1000:
        return 'Low'
    elif max_installs < 100000:
        return 'Medium'
    else:
        return 'High'

data['Maximum Installs'] = data['Maximum Installs'].apply(categorize_maximum_installs)

In [17]:
data['Minimum Installs']

0             Low
1          Medium
2             Low
3             Low
4             Low
            ...  
2312939      High
2312940       Low
2312941       Low
2312942    Medium
2312943       Low
Name: Minimum Installs, Length: 2278748, dtype: object

In [18]:
data['Minimum Installs']

0             Low
1          Medium
2             Low
3             Low
4             Low
            ...  
2312939      High
2312940       Low
2312941       Low
2312942    Medium
2312943       Low
Name: Minimum Installs, Length: 2278748, dtype: object

<h4 style='color:green'> ~ Content Rating </h4>

In [19]:
# to check all the ctaegory of content rating
print(data['Content Rating'].value_counts())

Content Rating
Everyone           1992976
Teen                192781
Mature 17+           59617
Everyone 10+         33088
Unrated                154
Adults only 18+        132
Name: count, dtype: int64


In [20]:
# converting to content rating to str
data['Content Rating'] = data['Content Rating'].astype(str)

# Create a mapping dictionary
content_rating_mapping = {
    'Everyone': 'Everyone',
    'Teen': 'Teenager',       
    'Mature 17+': 'Teenager',
    'Everyone 10+': '10+',     
    'Unrated': 'Unrated',
    'Adults only 18+': '18+'   
}
 
data['Content Rating'] = data['Content Rating'].map(content_rating_mapping)

In [21]:
data['Content Rating'].value_counts()

Content Rating
Everyone    1992976
Teenager     252398
10+           33088
Unrated         154
18+             132
Name: count, dtype: int64

<h4 style='color:green'> ~ Rating </h4>

In [22]:
print(data['Rating'].value_counts())

Rating
0.0    1048592
5.0     100103
4.2      87984
4.4      86298
4.3      83271
4.6      78296
4.5      76743
4.1      69718
4.0      67336
4.7      62199
4.8      61098
3.9      55679
3.8      53858
4.9      44521
3.7      43296
3.6      35984
3.5      31631
3.4      28903
3.3      22609
3.2      21483
3.0      17274
3.1      15539
2.8      12234
2.9      11518
2.6       9211
2.7       9199
2.5       6941
2.4       6049
2.3       5413
2.2       4951
2.0       3803
2.1       3674
1.8       2957
1.9       2587
1.7       1927
1.6       1644
1.5       1156
1.4       1011
1.0        712
1.3        579
1.2        531
1.1        236
Name: count, dtype: int64


In [23]:
# Convert the 'Rating' column to numeric, setting errors='coerce' to convert invalid entries to NaN
data['Rating'] = pd.to_numeric(data['Rating'], errors='coerce')

def categorize_rating(rating):
    if pd.isna(rating):
        return np.nan 
    elif rating >= 4.6:
        return "Excellent"
    elif rating >= 3.6:
        return "High"
    elif rating >= 2.1:
        return "Medium"
    else:
        return "Low"

# Apply the categorization function
data['Rating'] = data['Rating'].apply(categorize_rating)

data.dropna(subset=['Rating'], inplace=True)

In [24]:
data['Rating'].value_counts()

Rating
Low          1065735
High          660167
Excellent     346217
Medium        206629
Name: count, dtype: int64

<h4 style='color:green'> ~ In App Purchases </h4>

In [25]:
# Convert 'In App Purchases' to numeric, forcing errors to NaN for non-convertible values
data['In App Purchases'] = pd.to_numeric(data['In App Purchases'], errors='coerce')

# Function to categorize apps based on Free and In-App Purchases
def categorize_app(row):
    if row['Free']: 
        return "Free"
    elif pd.notna(row['In App Purchases']) and row['In App Purchases'] > 0: 
        return "In App Purchases"
    else: 
        return "Paid"

data['In App Purchases'] = data.apply(categorize_app, axis=1)

In [26]:
data['In App Purchases']

0          Free
1          Free
2          Free
3          Free
4          Free
           ... 
2312939    Free
2312940    Free
2312941    Free
2312942    Free
2312943    Free
Name: In App Purchases, Length: 2278748, dtype: object

<h4 style='color:green'> ~ Rating Count </h4>

In [27]:
data['Rating Count'].value_counts()

Rating Count
0.0          1048592
5.0            64275
6.0            54306
7.0            46827
8.0            40578
              ...   
9028.0             1
27999.0            1
1271735.0          1
29786.0            1
78351.0            1
Name: count, Length: 38482, dtype: int64

In [28]:
# Define bins and labels for categorizing rating counts, excluding "No Ratings"
bins = [0, 100, 1000, 10000, float('inf')]  # Bins starting from 0
labels = ['Low', 'Medium', 'High', 'Very High']

data['Rating Count'] = pd.cut(data['Rating Count'], bins=bins, labels=labels)

In [29]:
data['Rating Count']

0                NaN
1                Low
2                NaN
3                Low
4                NaN
             ...    
2312939    Very High
2312940          NaN
2312941          NaN
2312942          Low
2312943          Low
Name: Rating Count, Length: 2278748, dtype: category
Categories (4, object): ['Low' < 'Medium' < 'High' < 'Very High']

<h4 style='color:green'> ~ Price </h4>

In [30]:
data['Price'] = pd.to_numeric(data['Price'], errors='coerce')
# Define bins and labels
bins = [-1, 0, 50, 150, float('inf')]  # -1 for inclusive lower bounds
labels = ['Free', 'Low-cost', 'Medium-cost', 'High-cost']

data['Price'] = pd.cut(data['Price'], bins=bins, labels=labels)

In [31]:
data['Price'].value_counts()

Price
Free           2235163
Low-cost         43177
Medium-cost        320
High-cost           88
Name: count, dtype: int64

In [32]:
data

App Name       Category  \
0                                                  Gakondo      Adventure   
1                                      Ampere Battery Info          Tools   
2                                                   Vibook   Productivity   
3        Smart City Trichy Public Service Vehicles 17UC...  Communication   
4                                                  GROW.me          Tools   
...                                                    ...            ...   
2312939                                           大俠客—熱血歸來   Role Playing   
2312940                                         ORU Online      Education   
2312941                                     Data Structure      Education   
2312942                                        Devi Suktam  Music & Audio   
2312943                       Biliyor Musun - Sonsuz Yarış         Trivia   

            Rating Rating Count              Installs Minimum Installs  \
0              Low          NaN           0 to 10,000              Low   
1             High          Low           0 to 10,000           Medium   
2              Low          NaN           0 to 10,000              Low   
3        Excellent          Low           0 to 10,000              Low   
4              Low          NaN           0 to 10,000              Low   
...            ...          ...                   ...              ...   
2312939       High    Very High  100,001 to 1,000,000             High   
2312940        Low          NaN           0 to 10,000              Low   
2312941        Low          NaN           0 to 10,000              Low   
2312942     Medium          Low           0 to 10,000           Medium   
2312943  Excellent          Low           0 to 10,000              Low   

        Maximum Installs  Free Price Size Minimum Android  \
0                    Low  True  Free   10      5.0 to 8.0   
1                 Medium  True  Free    2      5.0 to 8.0   
2                    Low  True  Free    3       Below 5.0   
3                    Low  True  Free    1       Below 5.0   
4                    Low  True  Free    6       Below 5.0   
...                  ...   ...   ...  ...             ...   
2312939             High  True  Free   77       Below 5.0   
2312940              Low  True  Free   44       Below 5.0   
2312941              Low  True  Free   29      5.0 to 8.0   
2312942           Medium  True  Free   10      5.0 to 8.0   
2312943              Low  True  Free    5      5.0 to 8.0   

                              Developer Id      Released  Last Updated  \
0        Jean Confident Irénée NIYIZIBYOSE  Feb 26, 2020  Feb 26, 2020   
1                               Webserveis  May 21, 2020  May 06, 2021   
2                               Cabin Crew   Aug 9, 2019  Aug 19, 2019   
3                      Climate Smart Tech2  Sep 10, 2018  Oct 13, 2018   
4                    Rafal Milek-Horodyski  Feb 21, 2020  Nov 12, 2018   
...                                    ...           ...           ...   
2312939                         ALICE GAME       Unknown  Jun 01, 2021   
2312940                3Dream Studios, LLC  Jan 17, 2018  Feb 02, 2018   
2312941                          appoworld  Aug 19, 2018  Aug 19, 2018   
2312942                             Bhakti   Aug 1, 2016  May 05, 2021   
2312943                          Y Yazılım   Aug 9, 2019  Aug 19, 2019   

        Content Rating In App Purchases  Editors Choice         Scraped Time  
0             Everyone             Free           False  2021-06-15 20:19:35  
1             Everyone             Free           False  2021-06-15 20:19:35  
2             Everyone             Free           False  2021-06-15 20:19:35  
3             Everyone             Free           False  2021-06-15 20:19:35  
4             Everyone             Free           False  2021-06-15 20:19:35  
...                ...              ...             ...                  ...  
2312939       Teenager             Free           False  2021

In [33]:
data.to_csv('googleplay.csv', index=False, encoding='utf-8')